lib/utils/ignore.ts

```diff
@@ -182,3 +184,92 @@ export class FileIgnorer {
 		return !/^\.\.\//.test(path.posix.relative(path1, path2));
 	}
 }
+
+interface FileStats {
+	filePath: string;
+	relPath: string;
+	stats: fs.Stats;
+}
+
+/**
+ * Create a list of files (FileStats[]) for the filesystem subtree rooted at
+ * projectDir, listing each file with both a full path and a relative path,
+ * but excluding entries for directories themselves.
+ * @param projectDir Source directory (root of subtree to be listed)
+ * @param dir Used for recursive calls only (omit on first function call)
+ */
+async function listFiles(
+	projectDir: string,
+	dir: string = projectDir,
+): Promise<FileStats[]> {
+	const files: FileStats[] = [];
+	const dirEntries = await fs.readdir(dir);
+	await Promise.all(
+		dirEntries.map(async entry => {
+			const filePath = path.join(dir, entry);
+			const stats = await fs.stat(filePath);
+			if (stats.isDirectory()) {
+				files.push(...(await listFiles(projectDir, filePath)));
+			} else if (stats.isFile()) {
```

Comment:
Here `stats` will not refer to a symbolic link, because `fs.stat(filePath)` (unlike `fs.lstat`) returns stats for the linked file in case `filePath` is a symbolic link. (Unless perhaps it was a symbolic link pointing to another symbolic link... Not sure if `fs.stat` follows links recursively...) The `isFile()` test is to guard against other "special files" like Linux character devices, block devices and such.

Reply:
I have confirmed that `stats.isFile()` returns `true` even if the `filePath` is a symbolic link pointing to another symbolic link, pointing to a regular file. (Which suggests that `fs.stat()` follows symlinks recursively.)